File: src_py/tizenclaw/core/perception_engine.py
"""
TizenClaw Perception Engine — proactive device situation awareness.

Three-stage pipeline:
  1. DeviceProfiler: collects and analyzes system events into a ProfileSnapshot
  2. ContextFusionEngine: fuses snapshot data into a SituationAssessment with risk score
  3. ProactiveAdvisor: decides whether to suppress, notify, or evaluate based on assessment

Runs on a periodic analysis tick (default 30s) and publishes perception events.
"""
import asyncio
import json
import logging
import time
from enum import IntEnum
from typing import Dict, List, Any, Optional
from dataclasses import dataclass, field

logger = logging.getLogger(__name__)


# ── Data structures ────────────────────────────────────────────────


class SituationLevel(IntEnum):
    NORMAL = 0
    ADVISORY = 1
    WARNING = 2
    CRITICAL = 3


class AdvisoryAction(IntEnum):
    SUPPRESS = 0
    INJECT = 1   # inject context into next LLM call
    NOTIFY = 2   # send notification to user
    EVALUATE = 3 # trigger AgentCore evaluation


@dataclass
class ProfileSnapshot:
    battery_level: int = 100
    charging: bool = False
    battery_drain_rate: float = 0.0
    battery_health: str = "good"
    memory_trend: str = "stable"
    memory_warning_count: int = 0
    network_status: str = "connected"
    network_drop_count: int = 0
    foreground_app: str = ""
    timestamp: float = field(default_factory=time.time)


@dataclass
class SituationAssessment:
    level: SituationLevel = SituationLevel.NORMAL
    risk_score: float = 0.0
    summary: str = ""
    factors: List[str] = field(default_factory=list)
    suggestions: List[str] = field(default_factory=list)


@dataclass
class Advisory:
    action: AdvisoryAction = AdvisoryAction.SUPPRESS
    message: str = ""
    assessment: Optional[SituationAssessment] = None


# ── DeviceProfiler ─────────────────────────────────────────────────


class DeviceProfiler:
    """Collects system events and produces ProfileSnapshots."""

    def __init__(self):
        self._events: List[Dict[str, Any]] = []
        self._max_events = 500
        self._lock = asyncio.Lock()

    async def record_event(self, event):
        """Record a system event from the EventBus."""
        data = getattr(event, 'data', {}) or {}
        topic = getattr(event, 'topic', '') or ''
        async with self._lock:
            self._events.append({
                "topic": topic,
                "data": data,
                "timestamp": time.time(),
            })
            if len(self._events) > self._max_events:
                self._events = self._events[-self._max_events:]

    def get_event_count(self) -> int:
        return len(self._events)

    def analyze(self) -> ProfileSnapshot:
        """Analyze collected events into a snapshot."""
        snap = ProfileSnapshot()
        now = time.time()
        window = 300  # 5 minute analysis window

        for ev in reversed(self._events):
            if now - ev.get("timestamp", 0) > window:
                break
            topic = ev.get("topic", "")
            data = ev.get("data", {})

            if "battery" in topic:
                level = data.get("level", data.get("percent"))
                if level is not None:
                    snap.battery_level = int(level)
                charging = data.get("charging", data.get("is_charging"))
                if charging is not None:
                    snap.charging = bool(charging)

            elif "memory" in topic:
                level = data.get("level", "")
                if level:
                    snap.memory_trend = str(level)
                snap.memory_warning_count += 1

            elif "network" in topic:
                status = data.get("status", "")
                if status:
                    snap.network_status = status
                if "disconnect" in topic or status == "disconnected":
                    snap.network_drop_count += 1

            elif "app" in topic:
                app_id = data.get("app_id", "")
                if app_id:
                    snap.foreground_app = app_id

        # Classify battery health
        if snap.charging:
            snap.battery_health = "charging"
        elif snap.battery_level <= 5:
            snap.battery_health = "critical"
        elif snap.battery_level <= 15:
            snap.battery_health = "degrading"
        else:
            snap.battery_health = "good"

        return snap


# ── ContextFusionEngine ────────────────────────────────────────────


class ContextFusionEngine:
    """Fuses profile snapshot data into a SituationAssessment."""

    def fuse(self, snap: ProfileSnapshot,
             extra_context: Dict[str, Any] = None) -> SituationAssessment:
        """Evaluate risk from multiple dimensions."""
        assessment = SituationAssessment()
        risk = 0.0
        factors = []
        suggestions = []

        # ── Battery risk ──
        if not snap.charging:
            if snap.battery_level <= 5:
                risk += 0.4
                factors.append(f"배터리 위험: {snap.battery_level}%")
                suggestions.append("충전기를 연결하세요")
            elif snap.battery_level <= 15:
                risk += 0.25
                factors.append(f"배터리 부족: {snap.battery_level}%")
                suggestions.append("불필요한 앱을 종료하세요")
            elif snap.battery_level <= 30:
                risk += 0.1
                factors.append(f"배터리 주의: {snap.battery_level}%")

        # ── Memory risk ──
        if snap.memory_trend == "critical":
            risk += 0.3
            factors.append("메모리 위험 수준")
            suggestions.append("메모리를 많이 사용하는 프로세스를 종료하세요")
        elif snap.memory_warning_count >= 3:
            risk += 0.2
            factors.append(f"메모리 경고 {snap.memory_warning_count}회")

        # ── Network risk ──
        if snap.network_status == "disconnected":
            risk += 0.15
            factors.append("네트워크 연결 끊김")
            suggestions.append("Wi-Fi 또는 모바일 데이터 연결을 확인하세요")
        elif snap.network_drop_count >= 3:
            risk += 0.1
            factors.append(f"네트워크 불안정 (끊김 {snap.network_drop_count}회)")

        # ── Determine level ──
        risk = min(risk, 1.0)
        assessment.risk_score = risk
        assessment.factors = factors
        assessment.suggestions = suggestions

        if risk >= 0.7:
            assessment.level = SituationLevel.CRITICAL
            assessment.summary = "시스템이 위험 상태입니다"
        elif risk >= 0.4:
            assessment.level = SituationLevel.WARNING
            assessment.summary = "주의가 필요한 상태입니다"
        elif risk >= 0.2:
            assessment.level = SituationLevel.ADVISORY
            assessment.summary = "경미한 문제가 감지되었습니다"
        else:
            assessment.level = SituationLevel.NORMAL
            assessment.summary = "정상 상태입니다"

        return assessment

    @staticmethod
    def level_to_string(level: SituationLevel) -> str:
        return {
            SituationLevel.NORMAL: "normal",
            SituationLevel.ADVISORY: "advisory",
            SituationLevel.WARNING: "warning",
            SituationLevel.CRITICAL: "critical",
        }.get(level, "unknown")

    @staticmethod
    def to_json(assessment: SituationAssessment) -> Dict[str, Any]:
        return {
            "level": ContextFusionEngine.level_to_string(assessment.level),
            "level_num": int(assessment.level),
            "risk_score": assessment.risk_score,
            "summary": assessment.summary,
            "factors": assessment.factors,
            "suggestions": assessment.suggestions,
        }


# ── ProactiveAdvisor ───────────────────────────────────────────────


class ProactiveAdvisor:
    """Decides what action to take based on SituationAssessment."""

    def __init__(self, agent_core=None, event_bus=None):
        self._agent = agent_core
        self._event_bus = event_bus
        self._last_insight: Dict[str, Any] = {}
        self._last_notify_time: float = 0
        self._cooldown = 120  # seconds between notifications

    def evaluate(self, assessment: SituationAssessment) -> Advisory:
        """Determine action based on assessment level."""
        self._last_insight = ContextFusionEngine.to_json(assessment)
        advisory = Advisory(assessment=assessment)

        if assessment.level == SituationLevel.NORMAL:
            advisory.action = AdvisoryAction.SUPPRESS
            return advisory

        # Check cooldown
        now = time.time()
        if now - self._last_notify_time < self._cooldown:
            advisory.action = AdvisoryAction.INJECT
            advisory.message = assessment.summary
            return advisory

        if assessment.level == SituationLevel.CRITICAL:
            advisory.action = AdvisoryAction.EVALUATE
            advisory.message = (
                f"⚠️ {assessment.summary}\n"
                f"위험 요인: {', '.join(assessment.factors)}\n"
                f"권장 조치: {', '.join(assessment.suggestions)}"
            )
            self._last_notify_time = now
        elif assessment.level >= SituationLevel.WARNING:
            advisory.action = AdvisoryAction.NOTIFY
            advisory.message = (
                f"⚠️ {assessment.summary}\n"
                f"요인: {', '.join(assessment.factors)}"
            )
            self._last_notify_time = now
        else:
            advisory.action = AdvisoryAction.INJECT
            advisory.message = assessment.summary

        return advisory

    def get_last_insight(self) -> Dict[str, Any]:
        return self._last_insight


# ── PerceptionEngine (integration) ─────────────────────────────────


class PerceptionEngine:
    """
    Main perception engine that ties DeviceProfiler, ContextFusionEngine,
    and ProactiveAdvisor together with a periodic analysis loop.
    """

    ANALYSIS_INTERVAL = 30  # seconds

    def __init__(self, agent_core=None, event_bus=None,
                 notification_callback=None):
        self._agent = agent_core
        self._event_bus = event_bus
        self._notify = notification_callback
        self._profiler = DeviceProfiler()
        self._fusion = ContextFusionEngine()
        self._advisor = ProactiveAdvisor(agent_core, event_bus)
        self._running = False
        self._task: Optional[asyncio.Task] = None

    async def start(self):
        """Start the perception engine: subscribe to events + analysis loop."""
        self._running = True

        # Subscribe to all events for profiling
        if self._event_bus:
            await self._event_bus.subscribe("*", self._profiler.record_event)

        self._task = asyncio.create_task(self._analysis_loop())
        logger.info("PerceptionEngine: Started")

    async def stop(self):
        self._running = False
        if self._task:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
        logger.info("PerceptionEngine: Stopped")

    def is_running(self) -> bool:
        return self._running

    def get_insight(self) -> Dict[str, Any]:
        return self._advisor.get_last_insight()

    async def _analysis_loop(self):
        """Periodic analysis tick."""
        while self._running:
            try:
                await asyncio.sleep(self.ANALYSIS_INTERVAL)

                if self._profiler.get_event_count() == 0:
                    continue

                # Run analysis pipeline
                snapshot = self._profiler.analyze()
                assessment = self._fusion.fuse(snapshot)
                advisory = self._advisor.evaluate(assessment)

                # Publish perception event
                if self._event_bus and assessment.level >= SituationLevel.ADVISORY:
                    from tizenclaw.core.event_bus import Event
                    await self._event_bus.publish_fire_and_forget(Event(
                        topic="perception.situation_changed",
                        data=ContextFusionEngine.to_json(assessment),
                        source="perception",
                    ))

                # Take action
                if advisory.action == AdvisoryAction.NOTIFY and self._notify:
                    await self._notify(advisory.message)
                elif advisory.action == AdvisoryAction.EVALUATE:
                    if self._notify:
                        await self._notify(advisory.message)
                    if self._agent:
                        try:
                            prompt = (
                                f"Perception Engine 분석 결과:\n"
                                f"{advisory.message}\n\n"
                                f"현재 디바이스 상태를 확인하고 사용자에게 도움이 되는 조언을 해주세요."
                            )
                            await self._agent.process_prompt("perception", prompt)
                        except Exception as e:
                            logger.error(f"PerceptionEngine: Agent eval failed: {e}")

            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error(f"PerceptionEngine: Analysis error: {e}")

    def get_status(self) -> Dict[str, Any]:
        return {
            "running": self._running,
            "event_count": self._profiler.get_event_count(),
            "last_insight": self._advisor.get_last_insight(),
            "analysis_interval_sec": self.ANALYSIS_INTERVAL,
        }

File: src/tizenclaw/core/perception_engine.cc
/*
 * Copyright (c) 2026 Samsung Electronics Co., Ltd All Rights Reserved
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "perception_engine.hh"
#include <chrono>
#include <cstdio>
#include <fstream>
#include <mutex>

#include "../../common/logging.hh"

namespace tizenclaw {

PerceptionEngine::PerceptionEngine(
    AgentCore* agent,
    SystemContextProvider* context,
    ChannelRegistry* channels)
    : agent_(agent),
      context_(context),
      profiler_(std::make_unique<DeviceProfiler>()),
      fusion_(
          std::make_unique<ContextFusionEngine>()),
      advisor_(std::make_unique<ProactiveAdvisor>(
          agent, channels)) {}

PerceptionEngine::~PerceptionEngine() {
  Stop();
}

void PerceptionEngine::Start() {
  if (running_.load()) return;

  // Subscribe to all EventBus events
  subscription_id_ =
      EventBus::GetInstance().SubscribeAll(
          [this](const SystemEvent& event) {
            OnEvent(event);
          });

  // Start analysis loop
  running_.store(true);
  analysis_thread_ = std::thread(
      &PerceptionEngine::AnalysisLoop, this);

  LOG(INFO) << "PerceptionEngine started "
            << "(hybrid: event-driven + "
            << kAnalysisIntervalSec << "s tick)";
}

void PerceptionEngine::Stop() {
  if (!running_.load()) return;

  running_.store(false);

  // Wake analysis thread for clean shutdown
  {
    std::lock_guard<std::mutex> lock(wake_mutex_);
    wake_requested_.store(true);
  }
  wake_cv_.notify_one();

  if (analysis_thread_.joinable()) {
    analysis_thread_.join();
  }

  if (subscription_id_ >= 0) {
    EventBus::GetInstance().Unsubscribe(
        subscription_id_);
    subscription_id_ = -1;
  }

  LOG(INFO) << "PerceptionEngine stopped";
}

void PerceptionEngine::OnEvent(
    const SystemEvent& event) {
  // Skip our own synthetic events to avoid loop
  if (event.source == "perception") return;

  // Feed every event to DeviceProfiler
  profiler_->RecordEvent(event);

  // If the event is significant, wake the
  // analysis thread immediately (with debounce)
  if (IsSignificantEvent(event)) {
    auto now = std::chrono::duration_cast<
                   std::chrono::milliseconds>(
                   std::chrono::system_clock::now()
                       .time_since_epoch())
                   .count();
    auto last = last_event_driven_tick_.load();
    if ((now - last) >= kEventDrivenDebounceMs) {
      // Update debounce timestamp at point of
      // wake to prevent multiple rapid wakes
      last_event_driven_tick_.store(now);
      LOG(INFO) << "PerceptionEngine: significant "
                << "event '" << event.name
                << "' — triggering analysis";
      {
        std::lock_guard<std::mutex> lock(
            wake_mutex_);
        wake_requested_.store(true);
      }
      wake_cv_.notify_one();
    }
  }
}

bool PerceptionEngine::IsSignificantEvent(
    const SystemEvent& event) {
  // Battery critical (< 15%) or charging change
  if (event.type == EventType::kBatteryChanged) {
    if (event.data.contains("level")) {
      int level = event.data["level"].get<int>();
      if (level < 15) return true;
    }
    if (event.data.contains("charging")) {
      return true;  // Charging state change
    }
  }

  // Network disconnected
  if (event.type == EventType::kNetworkChanged) {
    if (event.name == "network.disconnected") {
      return true;
    }
  }

  // Memory warning
  if (event.type == EventType::kMemoryWarning) {
    return true;
  }

  // Display off (device going to sleep)
  if (event.type == EventType::kDisplayChanged) {
    if (event.data.contains("state")) {
      std::string state =
          event.data["state"].get<std::string>();
      if (state == "off") return true;
    }
  }

  return false;
}

void PerceptionEngine::AnalysisLoop() {
  LOG(INFO) << "PerceptionEngine analysis "
            << "loop started (hybrid mode)";

  while (running_.load()) {
    // Wait on CV with periodic timeout
    {
      std::unique_lock<std::mutex> lock(
          wake_mutex_);
      wake_cv_.wait_for(
          lock,
          std::chrono::seconds(
              kAnalysisIntervalSec),
          [this]() {
            return wake_requested_.load() ||
                   !running_.load();
          });
      wake_requested_.store(false);
    }

    if (!running_.load()) break;

    RunAnalysisTick();

    // Update debounce timestamp
    last_event_driven_tick_.store(
        std::chrono::duration_cast<
            std::chrono::milliseconds>(
            std::chrono::system_clock::now()
                .time_since_epoch())
            .count());
  }

  LOG(INFO) << "PerceptionEngine analysis "
            << "loop exited";
}

void PerceptionEngine::RunAnalysisTick() {
  nlohmann::json insight;

  // Step 2: Conditional profile and situation analysis
  if (profiler_->GetEventCount() >= kMinEventsForAnalysis) {
    auto profile = profiler_->Analyze();
    nlohmann::json device_state;
    if (context_) {
      auto ctx = context_->GetContextJson();
      if (ctx.contains("device")) {
        device_state = ctx["device"];
      }
    }
    auto assessment = fusion_->Fuse(profile, device_state);
    
    insight["situation"] = ContextFusionEngine::ToJson(assessment);
    insight["trends"] = {
        {"battery_drain_rate", profile.battery_drain_rate},
        {"battery_health", profile.battery_health},
        {"memory_trend", profile.memory_trend},
        {"network_stability",
         profile.network_drop_count == 0 ? "stable" : (profile.network_drop_count < 3 ? "degraded" : "unstable")},
        {"memory_warning_count", profile.memory_warning_count}
    };
    if (!profile.top_apps.empty()) insight["top_apps"] = profile.top_apps;
    if (!profile.foreground_app.empty()) insight["foreground_app"] = profile.foreground_app;
    if (!profile.anomalies.empty()) insight["anomalies"] = profile.anomalies;

    auto advisory = advisor_->Evaluate(assessment);
    advisor_->Execute(advisory);

    if (assessment.level != SituationLevel::kNormal) {
      LOG(INFO) << "PerceptionEngine: " << ContextFusionEngine::LevelToString(assessment.level)
                << " (risk=" << static_cast<int>(assessment.risk_score * 100)
                << "%, factors=" << assessment.factors.size() << ")";
    }
  }

  // Step 3: Inject built insight into SystemContextProvider
  if (context_) {
    context_->SetPerceptionInsight(insight);
  }
}

nlohmann::json PerceptionEngine::GetInsight()
    const {
  if (advisor_) {
    return advisor_->GetLastInsight();
  }
  return {};
}

nlohmann::json PerceptionEngine::GetStatus()
    const {
  nlohmann::json status;

  // Engine state
  status["engine"] = {
      {"running", running_.load()},
      {"mode", "hybrid_event_driven"},
      {"analysis_interval_sec",
       kAnalysisIntervalSec},
      {"event_debounce_ms",
       kEventDrivenDebounceMs},
      {"event_count",
       profiler_ ? profiler_->GetEventCount()
                 : 0}};

  // Current profile snapshot
  if (profiler_) {
    auto profile = profiler_->Analyze();
    status["profile"] = {
        {"battery_level", profile.battery_level},
        {"battery_drain_rate",
         std::round(profile.battery_drain_rate *
                    100) /
             100.0},
        {"battery_health", profile.battery_health},
        {"charging", profile.charging},
        {"memory_trend", profile.memory_trend},
        {"memory_warning_count",
         profile.memory_warning_count},
        {"network_status",
         profile.network_status},
        {"network_drop_count",
         profile.network_drop_count},
        {"foreground_app",
         profile.foreground_app},
        {"top_apps", profile.top_apps}};

    if (!profile.anomalies.empty()) {
      status["anomalies"] = profile.anomalies;
    }

    // Current situation assessment
    if (fusion_) {
      nlohmann::json device_state;
      if (context_) {
        auto ctx = context_->GetContextJson();
        if (ctx.contains("device")) {
          device_state = ctx["device"];
        }
      }
      auto assessment =
          fusion_->Fuse(profile, device_state);
      status["situation"] =
          ContextFusionEngine::ToJson(assessment);
    }
  }

  // Last insight from ProactiveAdvisor
  if (advisor_) {
    auto insight = advisor_->GetLastInsight();
    if (!insight.empty()) {
      status["last_insight"] = insight;
    }
  }

  return status;
}

}  // namespace tizenclaw
